### PR TITLE
Improve item moving and farming mode

### LINKED ISSUE #739: DIM can move recently transferred items if a destination is full.
When transferring items to a destination, one of the features of DIM is that it will make room if the destination is full.  DIM will select an item from criteria and move that item to the vault or an available character.

The issue is that DIM will move away an item I recently transfered to the character if it meets the criteria, even through I want that item to remain.  

We could approach this from a couple of angles.  
1.  We could create a queue of items which starts with the original items and then adds each transfered item.
2. We could memoize the transfers and add a timeout that adds them to the 'transferable' pool of items over time.
3. We could change the criteria and place an emphasis on moving an unlike item that is of the lowest light level.  Essentially try to find an item that is unlikely to be used.

In my case, I was moving engrams to a character that was full.  Moving the first engram was fine.  It moved a weapon out and put the engram on the character.  Second engram failed because DIM selected the engram I just moved and put it in the vault. :/

### LINKED ISSUE #743: Can't Move Item with Full Vault and Empty Type
I have no artifacts in the vault, and I'm trying to move one from character to character.  DIM reports that there is nothing that can be moved aside, but that isn't the case.  The vault is full, but something can be moved from consumables to another character.

screen shot 2016-08-03 at 9 46 11 pm

screen shot 2016-08-03 at 10 01 19 pm

screen shot 2016-08-03 at 10 02 00 pm

### LINKED ISSUE #1056: Glimmer consumables not consolidating in farming mode
When I'm farming with the farming mode turned on, my glimmer consumables are always moved to another character, even though I keep a stack of 200+ of each of them in the vault. The logic should direct them to be sent to any existing stacks. Related to this issue is that it doesn't consistently send them to the same character - I frequently end my farming sessions on my hunter with 10 House Banners having gone to my warlock and 8 to my titan, while also having the 200 that I started with in the vault.

http://imgur.com/a/319l7

### LINKED ISSUE #1141: Farming move-aside subtly broken
#1057 subtly broke how farming mode works - now, with a full vault but space on other characters, farming mode will refuse to transfer items off my primary character in order to make space. Before, it would move items out of the vault to make room. I'll fix it, but filing this issue to track it.

### PR DESCRIPTION
**Pull request by bhollis**
This is a decently major rework of how item moving and space availability is handled. There's a lot here because I got a bit carried away 😇, but I think it'll fix a lot of our funny edge cases. I also cleaned up farming mode quite a bit in order to restore its behavior to the original "stuff every nook and cranny" intention, while also attempting to get rid of the "thrashing" behavior where a full vault and full characters will cause items to start getting constantly moved back and forth (also known as "OMG I just got like five exotics... or not 😤").

There's enough changes here that it'll need a **bunch** of testing, and it's too late tonight for me to start running strikes. I've left in some debug logging to help understand what moves are being made or why.

Some things I've changed:

1. Moving an item into a full vault but an empty bucket (such as full General but the vault contains no Consumables) now works. Fixes #743.
1. Stacks of items are properly accounted for. They'll now combine as things are moved to make space - previously even a stack of 1 consumable would count as taking up the whole slot and would prevent a move of 2 more of that consumable. Fixes #1056 and #1056.
1. We now catch errors trying to move move-aside items and retry (adding the failed item to the excludes list). This is dangerous, and I haven't tested it happening yet, but it should handle some pathological cases like [unattuned artifacts](https://github.com/DestinyItemManager/DIM/issues/951). One scenario I'm worried about is an API error on all moves causing this to loop through every item trying to move it. 🚨
1. The aforementioned "thrashing" in farming mode is fixed. When farming mode can't proceed (because moving anything off the character would result in something else being moved back on, because you're out of space), we now show a friendly info message. This message is throttled to show up no more than once a minute.
1. Fixed the Beta bug where a full vault would prevent farming mode from moving things to other characters. Fixes #1141.
1. Not exactly a bugfix, but I've tweaked the move-aside logic to strongly prefer putting things on characters other than the original item's owner. This doesn't make it down into deep recursion chains (when we're towers-of-hanoi-ing our way out of very full inventory) but so far it seems to make it much easier to move a bunch of stuff off of a character without other things bouncing right back in. Sorta fixes #739, though I want to do more with this by recording the last manual move time per item.
2. I also tweaked the value function used in the move-aside function to strongly prefer putting engrams in the vault and not taking them out.
1. I've expanded farming mode's "make room" feature to include artifacts, materials, and consumables, since those tend to fill up and make annoying un-grabbables when I play too.
1. I've added a behavior where, when making space in the "General" category or in Materials/Consumables buckets, I'll choose to move aside an item that can be combined with another stack somewhere without increasing the total number of stacks. This trends towards consolidation and can help free up a full vault, as well as getting rid of stray stacks.
1. Added support for passing space reservations into the move command. This is step one towards implementing [make room for postmaster](https://github.com/DestinyItemManager/DIM/issues/502) and [empty slots in loadouts](https://github.com/DestinyItemManager/DIM/issues/224).
1. I swapped in "special ammo synth" and "primary ammo synth" instead of "motes of light" and "strange coins" for the farming mode quick gather buttons. They seemed more useful in the heat of battle.
1. The move aside target item and target store are now calculated together, rather than as separate functions. This is not only more efficient, but it prevents bugs where the item was chosen assuming it'd go to one place, but the second function chose a different destination.
1. I changed a lot of stuff to `let`/`const` and arrow functions. It's far from a complete sweep, but it made a bunch of stuff more concise and readable, such as the main `moveTo` method.

### CODE REVIEW

**Comment by bhollis** [comment]
Hmm looks like I broke something right at the end there while tweaking farming mode that transfers through the vault not quite right... 😅.

**Comment by Mulchman** [comment]
I can certainly test it out Tuesday evening!

**Comment by bhollis** [comment]
OK OK I fixed the bugs I found!

File: app/scripts/store/dimStoreBucket.directive.js [hunk 1]
```diff
@@ -164,6 +164,7 @@
       }
 
       promise = promise.then(function(moveAmount) {
+        console.log("User initiated move:", moveAmount, item.name, item.type, 'to', target.name, 'from', dimStoreService.getStore(item.owner).name);
```
Comment (kyleshay) [inline app/scripts/store/dimStoreBucket.directive.js]: intentional to keep in?

File: app/scripts/services/dimItemService.factory.js [hunk 2]
```diff
@@ -381,102 +373,215 @@
...
+        var moves = _.pairs(movesNeeded)
               .reverse()
-              .find(function(p) { return p[1] > 0; });
-        var source = dimStoreService.getStore(move[0]);
-        var moveAsideItem = chooseMoveAsideItem(source, item, moveContext);
-        var target = chooseMoveAsideTarget(source, moveAsideItem, moveContext);
-
-        if (!target || (!target.isVault && target.spaceLeftForItem(moveAsideItem) <= 0)) {
-          return $q.reject(new Error('There are too many \'' + (target.isVault ? moveAsideItem.bucket.sort : moveAsideItem.type) + '\' items in the ' + target.name + '.'));
+              .find(([_, moveAmount]) => moveAmount > 0);
```
Comment (kyleshay) [inline app/scripts/services/dimItemService.factory.js]: what is this? why is the underscore being passed in as an element in that array?

**Comment by Mulchman** [comment]
Didn't get to test it as much as I would have liked. Got to 107/108 weapons in the Vault, and engrams started getting put onto my other characters, but by that time it was too late and I quit for the night. (Running with both of the options disabled (`Move Uncommon/Green Items to Vault` & `Make room to pick up items by moving equipment`).)

File: app/scripts/store/dimStoreBucket.directive.js (hunk 1 above)
Comment (bhollis) [inline app/scripts/store/dimStoreBucket.directive.js]: Yeah, for now I'd like to keep all the logging in.

**Comment by bhollis** [comment]
I have found a few bugs in this, but I haven't had a lot of time. Give it another few weeks.

File: app/scripts/services/dimFarmingService.factory.js [hunk 3]
```diff
@@ -168,24 +181,23 @@
           var consolidateHashes = [
```
Comment (delphiactual) [inline app/scripts/services/dimFarmingService.factory.js]: Can't we make these user-defined, with a drag & drop, that persist. Like loadout builder has for locking slots.

**Comment by delphiactual** [comment]
Cannot move an equipped item to a different character, Error message "You cannot perform this action on an equipped item."

move-items-error

**Comment by bhollis** [comment]
Yup that's one of the bugs

-Ben

**Comment by bhollis** [comment]
Remaining issues I know about:

* [ ] Moving an equipped item fails.
* [ ] The recursive retry has no termination case, so if there's a problem w/ the API it'll try every item in your inventory before giving up.
* [ ] It isn't making space for consumables. This may be related to still-classified consumables in my inventory.

**Comment by bhollis** [comment]
OK, I also rewrote the logic we use to choose which item to equip when dequipping an item.

Now we keep looking past the first candidate, prefer replacing exotics with other exotics, and correctly handle The Life Exotic.

This fixes #1221, plus generally makes things work as expected more often.

**Comment by bhollis** [comment]
All right, those are the bugs I know about, all fixed. Let's get this into the beta and start testing for real. What do you think?

**Pull request merged**